scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/AdvancedInMemoryPskStore.java

```diff
@@ -0,0 +1,113 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Bosch.IO GmbH and others.
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *    Bosch.IO GmbH - initial creation
+ ******************************************************************************/
+package org.eclipse.californium.scandium.dtls.pskstore;
+
+import java.net.InetSocketAddress;
+
+import javax.crypto.SecretKey;
+
+import org.eclipse.californium.scandium.dtls.ConnectionId;
+import org.eclipse.californium.scandium.dtls.PskSecretResult;
+import org.eclipse.californium.scandium.dtls.PskSecretResultHandler;
+import org.eclipse.californium.scandium.dtls.PskPublicInformation;
+import org.eclipse.californium.scandium.dtls.cipher.PseudoRandomFunction;
+import org.eclipse.californium.scandium.dtls.cipher.ThreadLocalCryptoMap;
+import org.eclipse.californium.scandium.dtls.cipher.ThreadLocalCryptoMap.Factory;
+import org.eclipse.californium.scandium.dtls.cipher.ThreadLocalMac;
+import org.eclipse.californium.scandium.util.SecretUtil;
+import org.eclipse.californium.scandium.util.ServerNames;
+
+/**
+ * Simple in-memory example implementation of {@link AdvancedPskStore}.
+ * 
+ * Delegates calls to {@link PskStore}.
+ */
+public class AdvancedInMemoryPskStore implements AdvancedPskStore {
+
+	protected static final ThreadLocalCryptoMap<ThreadLocalMac> MAC = new ThreadLocalCryptoMap<>(
+			new Factory<ThreadLocalMac>() {
+
+				@Override
+				public ThreadLocalMac getInstance(String algorithm) {
+					return new ThreadLocalMac(algorithm);
+				}
+			});
+
+	protected final PskStore pskStore;
+	protected final boolean master;
+
+	/**
+	 * Create an advanced pskstore from {@link PskStore}.
+	 * 
+	 * @param pskStore psk store
+	 * @throws NullPointerException if store is {@code null}
+	 */
+	public AdvancedInMemoryPskStore(PskStore pskStore) {
+		if (pskStore == null) {
+			throw new NullPointerException("PSK store must not be null!");
+		}
+		this.pskStore = pskStore;
+		this.master = true;
+	}
+
+	/**
+	 * Create an advanced pskstore from {@link PskStore}.
+	 * 
+	 * @param pskStore psk store
+	 * @param master {@code true}, return master secret, {@code false}, return
+	 *            PSK secret key.
```

Comment:
You meant that getting "pure" secret key is only used for test ?

Reply:
The `AdvancedInMemoryPskStore` is mainly a example, used for testing. And the parameter is for testing the different behaviour returning the master secret or the private key.

Currently I work on some additional unit tests. I can try to move the "test parameters" also to the "AsyncInMemoryPskStore" and keep the "AdvancedInMemoryPskStore" more clean.

Reply:
You mean `AsyncInMemoryPskStore` (not `AdvancedInMemoryPskStore`) is mainly an example ? 

I think it's better to keep `AdvancedInMemoryPskStore` more clean because I didn't consider it as an exemple. I'm not sure but I don't exclude the possibility to used in Leshan with old PskStore API in a first time.

Reply:
OK, I try it.